### PR TITLE
Add Missing get/set Accessors and CUnit tests

### PR DESCRIPTION
**Pull request by hazen**
Addresses #14 and #10

### CODE REVIEW

File: src/include/messages/riak_delete.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ *********************************************************************/
+
+#ifndef _RIAK_DELETE_MESSAGE_H
+#define _RIAK_DELETE_MESSAGE_H
+
+typedef struct _riak_delete_options riak_delete_options;
+typedef struct _riak_delete_response riak_delete_response;
```
Comment (rzezeski) [inline src/include/messages/riak_delete.h:27]: Why the underscore for the actual struct? Why not just have struct and typedef named same thing?

**Comment by rzezeski** [comment]
It's going to take me a while to review this, but good news is it builds on OSX and passes `make check`.

File: src/include/messages/riak_delete.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+ * @param cfg Riak Configuration
+ * @param opt Delete options
+ * @param value Value of Write Quorum
+ */
+void
+riak_delete_options_set_w(riak_delete_options *opt,
+                          riak_uint32_t        value);
+/**
+ * @brief Set the Durable Write Quorum on a delete operation
+ * @param cfg Riak Configuration
```
Comment (rzezeski) [inline src/include/messages/riak_delete.h:89]: There is no `cfg` param.

File: src/include/messages/riak_delete.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ * @param opt Delete options
+ * @param value Value of Durable Write Quorum
+ */
+void
+riak_delete_options_set_dw(riak_delete_options *opt,
+                           riak_uint32_t        value);
+/**
+ * @brief Set the Primary Write Quorum on a delete operation
+ * @param cfg Riak Configuration
```
Comment (rzezeski) [inline src/include/messages/riak_delete.h:98]: No `cfg` param.

File: src/include/messages/riak_delete.h [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+ * @param opt Delete options
+ * @param value Value of Primary Write Quorum
+ */
+void
+riak_delete_options_set_pw(riak_delete_options *opt,
+                           riak_uint32_t        value);
+/**
+ * @brief Set the Time out on a delete operation
+ * @param cfg Riak Configuration
```
Comment (rzezeski) [inline src/include/messages/riak_delete.h:107]: No `cfg` param.

File: src/include/messages/riak_delete.h [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+ * @param opt Delete options
+ * @param value Value of Time Out (in ms)
+ */
+void
+riak_delete_options_set_timeout(riak_delete_options *opt,
+                                riak_uint32_t        value);
+/**
+ * @brief Allow a Sloppy Quorum on a delete operation
+ * @param cfg Riak Configuration
```
Comment (rzezeski) [inline src/include/messages/riak_delete.h:116]: No `cfg` param.

File: src/include/messages/riak_delete.h [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+ * @param opt Delete options
+ * @param value True or False
+ */
+void
+riak_delete_options_set_sloppy_quorum(riak_delete_options *opt,
+                                      riak_boolean_t       value);
+/**
+ * @brief Set the N Value on a delete operation
+ * @param cfg Riak Configuration
```
Comment (rzezeski) [inline src/include/messages/riak_delete.h:125]: No `cfg` param.

File: src/include/messages/riak_error.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+typedef struct _riak_error_response riak_error_response;
+
+/**
+ * @brief Free memory used by an error response
+ * @param cfg Riak Configuration
+ * @param resp Error structure to be freed
+ */
+void
+riak_free_error_response(riak_config         *cfg,
```
Comment (rzezeski) [inline src/include/messages/riak_error.h:34]: I notice other places put verb at end. E.g. `riak_delete_options_free`.

File: src/include/messages/riak_get.h [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+typedef struct _riak_get_response riak_get_response;
+typedef struct _riak_get_options riak_get_options;
+typedef void (*riak_get_response_callback)(riak_get_response *response, void *ptr);
+
+/**
+ * @brief Print a summary of a `riak_get_response`
+ * @param response Result from a Get request
+ * @param target Location of string to be formatted
+ * @param len Number of free bytes
```
Comment (rzezeski) [inline src/include/messages/riak_get.h:34]: The description for `len` is confusing. Free bytes in what? I'm assuming that len is the number of bytes malloc'd for `target`? If that is the case would "Size of target" or "Maximum number of bytes to write" make more sense?

File: src/include/messages/riak_get.h [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+riak_get_response_free(riak_config        *cfg,
+                       riak_get_response **resp);
+
+/**
+ * @brief Determine if Vector Clock is in a Get response
+ * @param response Riak Get Response
+ * @returns True if Vector Clock is returned
+ */
+riak_boolean_t
+riak_get_get_has_vclock(riak_get_response *response);
```
Comment (rzezeski) [inline src/include/messages/riak_get.h:57]: When would this function ever be used? If the response is not a not-found then there should always be a vector clock.

File: src/include/messages/riak_error.h (hunk 7 above)
Comment (hazen) [inline src/include/messages/riak_error.h:34]: Yes, that is one of my refactoring efforts: Consistent naming. The standard I agreed on was verb at the end. However, error and ping messages were not refactored in this particular PR, but can rename them here, too, if it's easier.

File: src/include/messages/riak_get.h (hunk 9 above)
Comment (rzezeski) [inline src/include/messages/riak_get.h:57]: Also, if we are going to have a function like this the suffix `get_has_vclock` seems strange. That, to me, would indicate it is fetching the field `has_vclock` but really it is determining if the `vclock` field exists. To me it should be called `riak_get_has_vclock`.

Comment (hazen) [inline src/include/messages/riak_get.h:57]: It would be used to access the "has_vclock" flag.  I would guess it might be a bigger, server-side question as to how valuable the information is.  As to the naming, sadly I have a message called "get" and a series of functions which are called "get" and "set".  My naming standard is `riak_<message>_get/set_<field>`.  I could rename the message `fetch`, perhaps, but I was trying to match the underlying PBC names.

File: src/include/messages/riak_error.h (hunk 7 above)
Comment (rzezeski) [inline src/include/messages/riak_error.h:34]: Probably not, this PR is big enough as it is :). Filing an issue would make sense.

File: src/include/messages/riak_get.h (hunk 9 above)
Comment (rzezeski) [inline src/include/messages/riak_get.h:57]: I should have looked at the struct first, sorry. I assumed that `has_clock` was derived by checking if `vclock` is `NULL`. Why not do that? Why an extra field?

Comment (hazen) [inline src/include/messages/riak_get.h:57]: Because that's what I get back from the server. For many of the fields (ints, bools and empty strings) there would be no easy way to determine if a value is populated or not.  Looks like that is the way in which protocol buffers handle the "optional" keyword, too.  The `protoc-c` compiler must be adding that `has_` field to clarify the presence of an optional field in a message.

https://github.com/basho/riak_pb/blob/develop/src/riak_kv.proto#L65

Comment (rzezeski) [inline src/include/messages/riak_get.h:57]: Ahhh, okay that makes more sense now. Sorry, I forget that it has to deal with the protobuff format underneath. I haven't done a macro pass on the C client yet so I'm failing to see bigger picture. Thanks for the patience.

File: src/include/messages/riak_put.h [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+
+/**
+ * @brief Print contents of a Riak Put Options to a string
+ * @param obj Object to print
+ * @param target Location to write formatted string
+ * @param len Number of bytes to write
+ * @return Number of bytes written
+ */
+int
+riak_put_options_print(riak_put_options *opt,
```
Comment (rzezeski) [inline src/include/messages/riak_put.h:75]: Shouldn't this return `riak_int32_t` since that's the type of the `len` arg?

File: src/include/messages/riak_serverinfo.h [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+ * @brief Free memory from response
+ * @param cfg Riak Configuration
+ * @param resp Server Info PBC Response
+ */
+void
+riak_serverinfo_response_free(riak_config               *cfg,
+                              riak_serverinfo_response **resp);
+
+/**
+ * @brief Determine if a node is present in the response
```
Comment (rzezeski) [inline src/include/messages/riak_serverinfo.h:50]: This description confused me. I would expect to be able to pass a node. What this function really does is determine if a node _field_ is present.

File: src/include/riak_bucketprops.h [hunk 12]
```diff
@@ -0,0 +1,794 @@
...
+riak_modfun_new(riak_config *cfg);
+
+/**
+ * @brief Create an Erlang module/function from another one
+ * @param cfg Riak Configuration
+ * @param from Source Riak Module/Function
+ * @returns Freshly baked Riak Mod/Fun
+ */
+riak_modfun*
+riak_modfun_deep_new(riak_config *cfg,
```
Comment (rzezeski) [inline src/include/riak_bucketprops.h:745]: Would this be better named copy?

File: src/internal/messages/riak_delete-internal.h [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ *********************************************************************/
+
+// Based on RpbDelReq
```
Comment (rzezeski) [inline src/internal/messages/riak_delete-internal.h:23]: Do we have to manually generate these structs? Shouldn't the protobuff lib do that for us? As it stands if the proto changes we now have to change the structs, function prototypes for API on those structs and implementations for those functions. That's 3 places.

File: src/include/messages/riak_put.h (hunk 10 above)
Comment (hazen) [inline src/include/messages/riak_put.h:75]: Yes. It would be more consistent.

File: src/include/messages/riak_serverinfo.h (hunk 11 above)
Comment (hazen) [inline src/include/messages/riak_serverinfo.h:50]: Sorry. Yes, technically the `node` field

File: src/include/riak_bucketprops.h (hunk 12 above)
Comment (hazen) [inline src/include/riak_bucketprops.h:745]: I need to find a better naming convention. I have `_new`, `_deep_new` and `_copy`.  Maybe I can kill one.  I agree the "deep" name is rather ugly.  When I first started I only did shallow copies to save on memory until I realized that was not always possible, so I added this new flavor.  Not limited to this module by any means.

File: src/internal/messages/riak_delete-internal.h (hunk 13 above)
Comment (hazen) [inline src/internal/messages/riak_delete-internal.h:23]: `protoc-c` does generate its own structs. We are creating our own to insulate the user from the lower-level protocol. As it stands, I could switch the transport to, say HTTP, and the user interface does not change. Same as if we switch to any zany new protocol of the month.  I was going to look longer term into code gening some of the these from the PBC headers.  Sadly there are lots of tiny difference, such as nested structs.  After talking with Dave, we decided it would be better to get a client out rather than creating more infrastructure at this point.  One thing we are working on is having the `riak_pb` project create our headers for us, too.

File: src/include/riak_bucketprops.h (hunk 12 above)
Comment (rzezeski) [inline src/include/riak_bucketprops.h:745]: The "deep" part is not what bothered me. I think the notion of deep vs shallow copy makes sense and is a common idiom. It's the `_new` suffix I found strange. I would have expected `_copy`.

File: src/internal/messages/riak_delete-internal.h (hunk 13 above)
Comment (rzezeski) [inline src/internal/messages/riak_delete-internal.h:23]: I totally agree on adding a user API that encapsulates the protocol. However, this current API is basically a one-for-one mapping to the PB protocol. That's what I find very strange about all this. I don't expect anything to change for this PR but I just have this hunch we can do better than simply wrapping the PB generated structs with our own that are basically identical. Just food for thought.

File: src/internal/messages/riak_reset_bucketprops-internal.h [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+// Based on RpbResetBucketReq
+struct _riak_reset_bucketprops_request
+{
+    riak_binary   *bucket;
+    riak_boolean_t has_type;
+    riak_binary   *type;
+};
+
+// Placeholder
+struct _riak_reset_bucketprops_response
```
Comment (rzezeski) [inline src/internal/messages/riak_reset_bucketprops-internal.h:32]: ?

File: src/internal/messages/riak_set_bucketprops-internal.h [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    riak_binary       *bucket;
+    riak_bucketprops *props;
+    riak_boolean_t     has_type;
+    riak_binary       *type;
+};
+
+// Placeholder
+struct _riak_set_bucketprops_response
+{
+// Empty
```
Comment (rzezeski) [inline src/internal/messages/riak_set_bucketprops-internal.h:35]: ?

File: src/internal/messages/riak_reset_bucketprops-internal.h (hunk 14 above)
Comment (hazen) [inline src/internal/messages/riak_reset_bucketprops-internal.h:32]: There's no payload to this message response; just the message code. The struct is a placeholder so we can pass something around as a response.  Might not be too useful, but there for completeness.

Comment (rzezeski) [inline src/internal/messages/riak_reset_bucketprops-internal.h:32]: No payload, so I would expect no corresponding struct. Just success or fail, with potential reason for failure. What's the point of passing around a struct the caller can do nothing with?

File: src/messages/riak_delete.c [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+    return ERIAK_OK;
+}
+
+void
+riak_delete_response_free(riak_config           *cfg,
+                          riak_delete_response **resp) {
+    riak_free(cfg, resp);
+}
+
+riak_boolean_t has_vclock;
```
Comment (rzezeski) [inline src/messages/riak_delete.c:120]: Is it just me or are these static variables not used anywhere? I don't see them used here or extern'd anywhere.

File: src/internal/messages/riak_reset_bucketprops-internal.h (hunk 14 above)
Comment (hazen) [inline src/internal/messages/riak_reset_bucketprops-internal.h:32]: The decoder, in this case `riak_reset_bucketprops_response_decode()`, is called for each message.  Currently they expect a response message.  Probably could have just made it a void*, but when I started I did not realize we would have any empty messages.  Good thought for cleanup.

**Comment by rzezeski** [comment]
TL;DR: +1 to merge.

Okay, as 2.0 approaches it's getting obvious that I just wont have time to properly finish reviewing this PR. It's just too big and I'm too new to both C and this client to review this with any speed. In fact, I think I should play with the client a bit more before any future reviews so I can get a better feeling for it. It would also be nice if future patches aren't this large. Many of my comment above still stand but I'll leave it up to you if you want to act on them. Since we are just trying to get something out and this is blocking other PRs I'm fine with seeing it merged and implementation details can be hashed out later when there is more time to discuss.

**Comment by hazen** [comment]
Sounds good, @rzezeski.  Thanks for all the effort you already put in.  I'll address what I can, but a deeper dive sounds like a great next step once I have the initial version released.

File: src/messages/riak_delete.c (hunk 16 above)
Comment (hazen) [inline src/messages/riak_delete.c:120]: Nuked
